File: LeetCode/0005_longest-palindromic-substring/20220519.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        max_length = 1
        ans = s[0]

        for i in range(len(s)):
            length0, s0 = self.f(s, l=i, r=i)
            if length0 > max_length:
                max_length = length0
                ans = s0
                
            length1, s1 = self.f(s, l=i, r=i+1)
            if length1 > max_length:
                max_length = length1
                ans = s1
                
        return ans
    
    def f(self, s: str, l: int, r: int) -> tuple[int, str]:
        max_length = 0
        ans = ""
        
        while l >= 0 and r < len(s) and s[l] == s[r]:
            max_length = r - l + 1
            ans = s[l:r+1]
            
            l -= 1
            r += 1
        
        return max_length, ans
        

File: LeetCode/Medium/0005_longest-palindromic-substring/20220519.kt
class Solution {
    fun longestPalindrome(s: String): String {
        var ans = s.substring(0, 1)
        
        (0..s.length - 1).forEach {
            val s0 = f(s, it, it)
            if (s0.length > ans.length) {
                ans = s0
            }
            
            val s1 = f(s, it, it + 1)
            if (s1.length > ans.length) {
                ans = s1
            }
        }
        
        return ans
    }
    
    fun f(s: String, l: Int, r: Int): String {
        var offset = 0
        
        while (l - offset >= 0 && r + offset < s.length && s[l - offset] == s[r + offset]) {
            offset += 1
        }
        
        offset -= 1
        
        return if (offset >= 0) s.substring(l - offset, r + offset + 1) else ""
    }
}